### ISSUE TITLE
[clang][c23] macro + non-void cast incorrectly suppresses "nodiscard" warning

### ISSUE THREAD

**Issue by poconn** [llvm/llvm-project#104391]
clang-18 compiles the following code warning-free at `-Weverything`, but should warn because the result of `f()` is essentially discarded (aside from the cast, but it would be to void if the result was supposed to be ignorable):
```c
[[nodiscard]] int f();

void g() {
#define M (unsigned int)f()
  M;
}
```
https://godbolt.org/z/9sdcY3eof

It does correctly warn when compiling as C++ (even with the C-style cast).
